pkg/server/fsm.go

```diff
@@ -206,6 +207,150 @@ func (s *adminStateRaw) Store(state adminState) {
 	s.val.Store(int32(state))
 }
 
+func initializeConn(fsm *fsm, conn net.Conn) {
+	conn.SetWriteDeadline(time.Now().Add(time.Second))
+
+	fsm.lock.Lock()
+	if err := setPeerConnTTL(fsm, conn); err != nil {
+		fsm.logger.Warn("cannot set TTL",
+			slog.String("State", fsm.state.String()),
+			slog.String("Error", err.Error()))
+	}
+	if err := setPeerConnMSS(fsm, conn); err != nil {
+		fsm.logger.Warn("cannot set MSS",
+			slog.String("State", fsm.state.String()),
+			slog.String("Error", err.Error()))
+	}
+	fsm.lock.Unlock()
+}
+
+type outgoingConn struct {
+	conn net.Conn
+	open *bgp.BGPMessage
+}
+
+type outgoingConnManager struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	fsm    *fsm
+}
+
+func newOutGoingConnManager(ctx context.Context, fsm *fsm) *outgoingConnManager {
+	cctx, cancel := context.WithCancel(ctx)
+	ocm := &outgoingConnManager{
+		ctx:    cctx,
+		cancel: cancel,
+		fsm:    fsm,
+	}
+	ocm.wg.Add(1)
+	go ocm.run(fsm.outgoingConnCh)
+
+	return ocm
+}
+
+func (ocm *outgoingConnManager) run(ch chan<- outgoingConn) {
+	defer func() {
+		ocm.wg.Done()
+		ocm.cancel()
+	}()
+
+	fsm := ocm.fsm
+	fsm.lock.Lock()
+	isPassive := fsm.pConf.Transport.Config.PassiveMode
+	fsm.lock.Unlock()
+
+	if isPassive {
+		return
+	}
+
+	state := bgp.BGP_FSM_CONNECT
+	var conn net.Conn
+	for {
+		switch state {
+		case bgp.BGP_FSM_CONNECT:
+			connCh := make(chan net.Conn, 1)
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go ocm.fsm.h.connectLoop(ocm.ctx, connCh, &wg)
+			select {
+			case <-ocm.ctx.Done():
+				wg.Wait()
+				return
+			case conn = <-connCh:
+				wg.Wait()
```

Comment:
Hey @fujita, do you really need to spawn a goroutine for `ocm.fsm.h.connectLoop`
I see that you are doing `wg.Wait()` as the first step of all the cases. So, the only consequence is that if the connection is received before the context is Done during the run of `connectLoop`, you would still proceed with the connection.

In my view, it would be even better to check the context after the run of `connectLoop` entirely, don't you think?

Reply:
Ah, indeed. You are right. We don't need to use goroutine here (and connCh channel) because context can cancel `connectLoop()` properly.

I'll update the code.